sdks/go/node/api/client/embedLocalFilesAndDirs.go

```diff
@@ -0,0 +1,73 @@
+package client
+
+import (
+	"fmt"
+	"github.com/opctl/opctl/sdks/go/model"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+)
+
+//embedLocalFilesAndDirs mutates args by embedding any referenced file &/or dirs as objects.
+// This makes the args location independent and therefore runnable on a remote node.
+// note1: this loads referenced files/dirs into memory so can easily exhaust available memory.
+// note2: this approach is primitive; we will need to support de-dupe and chunking in the future.
+func embedLocalFilesAndDirs(
+	args map[string]*model.Value,
+) error {
+	for key, val := range args {
+		var fileOrDirPath string
+		if nil != val.Dir {
+			fileOrDirPath = *val.Dir
+		} else if nil != val.File {
+			fileOrDirPath = *val.File
+		} else {
+			continue
+		}
+
+		fileObj, err := fileOrDirPathToObject(fileOrDirPath)
+		if nil != err {
+			return err
+		}
+
+		args[key] = &model.Value{Object: &fileObj}
+	}
+	return nil
+}
+
+// fileOrDirPathToObject recursively serializes a file or directory to an object
+func fileOrDirPathToObject(
+	path string,
+) (map[string]interface{}, error) {
+	info, err := os.Stat(path)
+	if nil != err {
+		return nil, err
+	}
+
+	if !info.IsDir() {
+		body, err := ioutil.ReadFile(path)
```

Comment:
I think it would be worth adding a max size check here (over the entire recursive operation), given the known potentially memory issues and the potential negative impact on the remote node if it gets sent a large amount of information.

Reply:
The max seems arbitrary here since the remote node will have arbitrary ram. Do you have more details about your thoughts? This is basically meant to be an incremental step towards the proper solution which would leverage content defined chunking (CDC) and dedupe via further enhancements to our data API. In practice the code here is used today by at least one client I know of and so getting it into the SDK so that future enhancements can be centrally rolled out was the main goal.

Reply:
This happens locally, though, right?

Just add a check that we're not reading more than, say, 200 mb and bail if that happens. It'll catch obvious issues like someone trying to mount their whole disk or a giant cache tree or something.

Reply:
The thing is; I know clients doing this today that in practice are sending upwards of 2GB.

Reply:
I *hope* this won't be a thing for long. It's a primitive temporary solution but since it's the only possible solution today, it's what clients are doing. We need to do more work on the universal data API to support doing this the right way. I think if we set it to 2GB it would suffice for now; it feels arbitrary but I'm onboard if you'd like to see it.